connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanner.java

```diff
@@ -118,35 +120,80 @@ private void loadJdbcDrivers(final ClassLoader loader) {
     }
 
     @SuppressWarnings({"rawtypes", "unchecked"})
-    protected <T> PluginDesc<T> pluginDesc(Class<? extends T> plugin, String version, ClassLoader loader) {
-        return new PluginDesc(plugin, version, loader);
+    protected <T> PluginDesc<T> pluginDesc(Class<? extends T> plugin, String version, PluginSource source) {
+        return new PluginDesc(plugin, version, source.loader());
     }
 
     @SuppressWarnings("unchecked")
-    protected <T> SortedSet<PluginDesc<T>> getServiceLoaderPluginDesc(Class<T> klass, ClassLoader loader) {
+    protected <T> SortedSet<PluginDesc<T>> getServiceLoaderPluginDesc(Class<T> klass, PluginSource source) {
         SortedSet<PluginDesc<T>> result = new TreeSet<>();
-        ServiceLoader<T> serviceLoader = ServiceLoader.load(klass, loader);
-        for (Iterator<T> iterator = serviceLoader.iterator(); iterator.hasNext(); ) {
-            try (LoaderSwap loaderSwap = withClassLoader(loader)) {
+        ServiceLoader<T> serviceLoader = handleLinkageError(klass, source, () -> ServiceLoader.load(klass, source.loader()));
```

Comment:
We added the `handleLinkageError` method because it was possible that calls to `Iterator::hasNext` and `Iterator::next` might skip over bad plugins, even if they threw a `LinkageError`. Do we have any reason to expect the same with `ServiceLoader::load` or `ServiceLoader::iterator`? If not, IMO we should leave these calls unwrapped.

Reply:
In the case where an Implementation of ServiceLoader may eagerly-evaluate a first provider and cause a LinkageError to appear, I wanted to re-use the handleLinkageError logging.

I don't think that the retries for these operations would be effective, and that the later retries would probably be wasteful. However, I thought re-using the same function in both places was simpler than customizing the behavior of each call. Since this code is (probably never) going to be exercised, I tried to keep it as simple as possible.

Do you think that we should log errors from load and iterator? I can catch the error or just let it propagate to the caller.

Reply:
I think that given the low likelihood of ever hitting this error when invoking `ServiceLoader::load` or `ServiceLoader::iterator` (neither is annotated with any kind of `throws` declaration, the iterator is specifically called out as being lazy-loading, and I'm not aware of any JDK bug reports that would lead to this kind of behavior), it's better to throw errors to the caller.

Reply:
I've removed these usages, so the error propagates normally.